### PR TITLE
compilertest (UI test): Support custom normalization.

### LINKED ISSUE #42434: Add wildcard support for UI tests
Some diagnostics have different output in different architectures, which [makes it harder to write arch agnostic UI tests](https://github.com/rust-lang/rust/pull/42304#issuecomment-305862331). This has cropped up multiple times. Personally I've either reworked the test or changed it to a compile-fail test, but given that [we'd like to consolidate on UI tests exclusively](https://github.com/rust-lang/rust/issues/40882), UI tests should have a way to avoid this problem.

### PR DESCRIPTION
**Pull request by kennytm**
Closes #42434. 

Adds this header for UI tests:

```rust
// normalize-stderr-32bit: "fn() (32 bits)" -> "fn() ($PTR bits)"
```

It will normalize the `stderr` output on 32-bit platforms, by replacing all instances of `fn() (32 bits)` by `fn() ($PTR bits)`. 

Extends the UI tests in #42304 and #41968 to 32-bit targets.

r? @nikomatsakis

### CODE REVIEW

**Comment by alexcrichton** [comment]
Looks like there may be some travis failure?

**Comment by kennytm** [comment]
@alexcrichton Travis is [fine](https://travis-ci.org/rust-lang/rust/jobs/250652733). `dist-aarch64-linux` [spuriously failed](https://travis-ci.org/rust-lang/rust/jobs/250652688) because it was unable to download `stamp`. (Ideally Travis-PR should only run one job but that's not possible AFAIK).

**Comment by nikomatsakis** [comment]
Ooh, exciting :)

File: src/test/ui/transmute/main.rs [hunk 1]
```diff
@@ -8,10 +8,10 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-// ignore-x86
-// ignore-arm
-// ignore-emscripten
-// ignore 32-bit platforms (test output is different)
+// normalize-stderr-32bit: "&str (64 bits)" -> "&str ($STR bits)"
+// normalize-stderr-64bit: "&str (128 bits)" -> "&str ($STR bits)"
```
Comment (nikomatsakis) [inline src/test/ui/transmute/main.rs]: nice

**Review by nikomatsakis — APPROVED** [review]
This looks great. Could you however update the [UI README file](https://github.com/rust-lang/rust/blob/master/src/test/ui/README.md) to describe how it works?

It's possible that this file should be merge with the more general [COMPILER_TESTS.md README](https://github.com/rust-lang/rust/blob/master/src/test/COMPILER_TESTS.md), as well.

**Comment by kennytm** [comment]
@nikomatsakis The two READMEs are merged in 7efb58c.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@kennytm nice

**Comment by kennytm** [comment]
[`dist-powerpc-linux` stalled during `make prepare`](https://travis-ci.org/rust-lang/rust/jobs/251410780). Spurious.

<details>

```
Attempting with retry: make prepare
[00:08:34] downloading https://static.rust-lang.org/dist/2017-06-15/rust-std-beta-x86_64-unknown-linux-gnu.tar.gz
######################################################################## 100.0%
[00:10:11] extracting /checkout/obj/build/cache/2017-06-15/rust-std-beta-x86_64-unknown-linux-gnu.tar.gz
[00:10:11] downloading https://static.rust-lang.org/dist/2017-06-15/rustc-beta-x86_64-unknown-linux-gnu.tar.gz
######################################################################## 100.0%
[00:11:40] extracting /checkout/obj/build/cache/2017-06-15/rustc-beta-x86_64-unknown-linux-gnu.tar.gz
[00:11:40] downloading https://static.rust-lang.org/dist/2017-06-15/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
######################################################################## 100.0%
[00:11:43] extracting /checkout/obj/build/cache/2017-06-15/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
[00:11:43]     Updating registry `https://github.com/rust-lang/crates.io-index`
[00:17:24]     Updating git repository `https://github.com/rust-lang/cargo`

No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

</details>

**Comment by kennytm** [comment]
Legit. Well transmuting a 32-bit pointer to `u32` is certainly valid on 32-bit platforms 😂 Going to change the test to use `[u8; 3]`.

**Comment by kennytm** [comment]
@nikomatsakis [Updated test](https://github.com/rust-lang/rust/pull/43083/commits/ef74420be4a2286feb9910e0516879c2c8c1671b#diff-097b1387981cb00dca07dfb96988ad37). Now only one UI test left (`ui/transmute/main.rs`) really needs normalization.

**Comment by kennytm** [comment]
Hmm the error message is changed? Going to rebase and update the UI test.

```
[01:11:38] -error[E0512]: transmute called with types of different sizes
[01:11:38] +error[E0591]: can't transmute zero-sized type
[01:11:38]    --> $DIR/transmute-from-fn-item-types-error.rs:14:13
[01:11:38]     |
[01:11:38]  14 |     let i = mem::transmute(bar);
[01:11:38]     |             ^^^^^^^^^^^^^^
[01:11:38]     |
[01:11:38] -   = note: source type: unsafe fn() {bar} (0 bits)
[01:11:38] -   = note: target type: i32 (32 bits)
[01:11:38] +   = note: source type: unsafe fn() {bar}
[01:11:38] +   = note: target type: i32
[01:11:38] +   = help: cast with `as` to a pointer instead
```

**Comment by kennytm** [comment]
@nikomatsakis Updated test.

**Pull request merged**